File: validator-rust/src/state/storage.rs
use std::{num::NonZeroU64, path::Path};

use anyhow::{Context as _, Result};
use rusqlite::{Connection, params};

use super::ValidatorState;

const DEFAULT_STATE_HISTORY: NonZeroU64 = NonZeroU64::new(5).unwrap();
const DEFAULT_STORAGE_FILE: &str = ":memory:";

pub struct Storage {
    conn: Connection,
    history: NonZeroU64,
}

impl Storage {
    pub fn open(path: Option<&Path>, history: Option<NonZeroU64>) -> Result<Self> {
        let conn = Connection::open(path.unwrap_or(Path::new(DEFAULT_STORAGE_FILE)))?;
        conn.execute_batch(
            "CREATE TABLE IF NOT EXISTS validator_state (
                block_number INTEGER PRIMARY KEY,
                state_json   TEXT NOT NULL
            );",
        )?;
        Ok(Self {
            conn,
            history: history.unwrap_or(DEFAULT_STATE_HISTORY),
        })
    }

    pub fn save(&self, block_number: u64, state: &ValidatorState) -> Result<()> {
        let json = serde_json::to_string(state)?;
        self.conn.execute(
            "INSERT INTO validator_state (block_number, state_json) VALUES (?1, ?2)",
            params![i64::try_from(block_number)?, json],
        )?;
        if let Some(cleanup) = block_number.checked_sub(self.history.get()) {
            self.conn.execute(
                "DELETE FROM validator_state WHERE block_number <= ?1",
                params![i64::try_from(cleanup)?],
            )?;
        }
        Ok(())
    }

    pub fn load_latest(&self) -> Result<Option<ValidatorState>> {
        let result = self.conn.query_row(
            "SELECT state_json FROM validator_state ORDER BY block_number DESC LIMIT 1",
            [],
            |row| row.get::<_, String>(0),
        );
        match result {
            Ok(json) => Ok(Some(
                serde_json::from_str(&json).context("failed to deserialize validator state")?,
            )),
            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
            Err(e) => Err(e.into()),
        }
    }
}

File: validator-go/storage/storage.go
package storage

import (
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"

	_ "modernc.org/sqlite" // register the "sqlite" driver

	"github.com/safe-research/safenet/validator-go/state"
)

const schema = `
CREATE TABLE IF NOT EXISTS validator_state (
    block_number INTEGER NOT NULL PRIMARY KEY,
    state        TEXT    NOT NULL
);`

// Storage persists validator state snapshots in a SQLite database, retaining
// at most stateHistory entries. Zero means retain all.
type Storage struct {
	db           *sql.DB
	stateHistory uint
}

// Open opens (or creates) a SQLite database at path. An empty path uses an
// in-memory database. stateHistory controls how many snapshots are retained;
// zero keeps all snapshots.
func Open(path string, stateHistory uint) (*Storage, error) {
	dsn := path
	if path == "" {
		dsn = ":memory:"
	}
	db, err := sql.Open("sqlite", dsn)
	if err != nil {
		return nil, fmt.Errorf("open database: %w", err)
	}
	if _, err := db.Exec(schema); err != nil {
		db.Close()
		return nil, fmt.Errorf("create schema: %w", err)
	}
	return &Storage{db: db, stateHistory: stateHistory}, nil
}

// Save persists st at the given blockNumber. If stateHistory > 0, all but the
// most recent stateHistory rows are pruned within the same transaction.
func (s *Storage) Save(blockNumber uint64, st *state.State) error {
	data, err := json.Marshal(st)
	if err != nil {
		return fmt.Errorf("marshal state: %w", err)
	}

	tx, err := s.db.Begin()
	if err != nil {
		return fmt.Errorf("begin transaction: %w", err)
	}
	defer tx.Rollback() //nolint:errcheck

	if _, err := tx.Exec(
		"INSERT OR REPLACE INTO validator_state (block_number, state) VALUES (?, ?)",
		blockNumber, string(data),
	); err != nil {
		return fmt.Errorf("insert state: %w", err)
	}

	if s.stateHistory > 0 {
		if _, err := tx.Exec(`
			DELETE FROM validator_state
			WHERE block_number NOT IN (
				SELECT block_number FROM validator_state
				ORDER BY block_number DESC
				LIMIT ?
			)`, s.stateHistory,
		); err != nil {
			return fmt.Errorf("prune state: %w", err)
		}
	}

	return tx.Commit()
}

// LoadLatest returns the block number and state of the most recently saved
// snapshot. Returns (0, nil, nil) when the database is empty.
func (s *Storage) LoadLatest() (uint64, *state.State, error) {
	var blockNumber uint64
	var raw string

	err := s.db.QueryRow(
		"SELECT block_number, state FROM validator_state ORDER BY block_number DESC LIMIT 1",
	).Scan(&blockNumber, &raw)
	if errors.Is(err, sql.ErrNoRows) {
		return 0, nil, nil
	}
	if err != nil {
		return 0, nil, fmt.Errorf("query state: %w", err)
	}

	var st state.State
	if err := json.Unmarshal([]byte(raw), &st); err != nil {
		return 0, nil, fmt.Errorf("unmarshal state at block %d: %w", blockNumber, err)
	}
	return blockNumber, &st, nil
}

// Close releases all resources held by the storage.
func (s *Storage) Close() error {
	return s.db.Close()
}
